transaction-view/src/transaction_view.rs

```diff
@@ -14,19 +15,28 @@ use {
 /// about the layout of the serialized transaction.
 /// The owned `data` is abstracted through the `TransactionData` trait,
 /// so that different containers for the serialized transaction can be used.
-pub struct TransactionView<D: TransactionData> {
-    data: D,
-    meta: TransactionMeta,
+pub struct TransactionView<const SANITIZED: bool, D: TransactionData> {
```

Comment:
@tao-stones @alessandrod

What are your thoughts on this approach wrt encoding whether we sanitzed in the type itself?

Reply:
Basically I'm weighing whether or not we should even expose an unsanitized type at all, or if that should be some internal that's not exposed.

I think because the sanitize checks can be relatively expensive, it makes sense to allow calling code to opt-out of doing them until necessary.

Reply:
wrt the use of const generic parameter, my first thought is if it's aiming for having type-level compile-time guarantee, or if a runtime `is_sanitized` flag or enum would suffice? My hesitation towards const generic parameter is the code complexity, and in often cases, code bloat it adds. Just a quick thought, happy to discuss more.

I agree that exposing unsanitized type provides flexibility, user should not be forced to sanitize transaction before viewing it.

Reply:
I think we should have some compile-time guarantee tbh; we want to be sure we aren't processing transactions that don't pass those checks.

In terms of complexity I'm not sure it adds too much, I probably should make these aliases pub:
```rust
// alias for convenience
pub type UnsanitizedTransactionView<D> = TransactionView<false, D>;
pub type SanitizedTransactionView<D> = TransactionView<true, D>;
```

What I attempted to achieve with the const generic is the type safety separation (good) of `VersionedTransaction` vs `SanitizedVersionedTransaction`, but without having the mess of different member functions or calling fns on fields that make using `SanitizedVersionedTransaction` a pain in the ass to work with.

Using const generic, both have exactly the same fns and data, it's just one has gone through neccessary checks vs not.

Reply:
In terms of transitioning through the stages of our pipeline.

1. receive packet
2. parse -> `UnsanitizedTransactionView`
3. (possible filtering on priority, etc)
4. sanitize -> `SanitizedTransactionView`
5. static_meta -> `RuntimeTransaction<SanitizedTransactionView<D>>`
6. dynamic_meta -> `RuntimeTransaction<Resolved(?)TransactionView<D>>` // don't know about naming here yet, but basically wrapper of `SanitizedTransactionView` plus `LoadedAddresses`

Reply:
However, since I'm not sure we will actually do 3 - it may just be better to always do the sanitize checks.
wdyt?

Reply:
> Using const generic, both have exactly the same fns and data, it's just one has gone through neccessary checks vs not.

I don't think we actually want exactly the same functions and data for sanitized vs unsanitized do we? The current separation of `Sanitized` types and their unsanitized counterparts allowed us to add certain functions which we knew were only safe if the transaction was sanitized.

It looks like you've added some sanitization to `TransactionMeta` regardless of whether the sanitized or unsanitized view is used which allows us to know a transaction has at least one signature, for example. This would allow us to add a `signature` method to both sanitized and unsanitized views, which is nice..

But what about methods like `get_durable_nonce` or `program_instructions_iter` which rely on sanitized inputs? Are you not planning to add methods on the `SanitizedTransactionView` for those? If you are, I don't think it's a big difference to use separate structs vs this const generic approach.

Reply:
The const generic doesn't restrict us in this manner; I did not mean to imply they should have exactly the same functions - but that the `Sanitized` version should have access to **at least** the same functions as the `Unsanitized`. and that they should not be less convenient to call (which is my issue with `SanitizedVersionedTransaction`).

We can still add specific behavior (if we find it necessary) to the sanitized version like this:
```rust
// Additional methods that rely on sanitization checks for correctness.
impl<D: TransactionData> TransactionView<true, D> {
    pub fn program_instructions_iter<'a>(
        &'a self,
    ) -> impl Iterator<Item = (&'a Pubkey, SVMInstruction<'a>)> {
        self.instructions_iter().map(|ix| {
            (
                &self.static_account_keys()[usize::from(ix.program_id_index)],
                ix,
            )
        })
    }
}
```

You mentioned a few examples, so I'll just comment on those here as well.

- `get_durable_nonce` shouldn't be a member fn imo. It will need to become generic over `SVMTransaction` or similar trait so we can abstract the processing pipeline. And it can be simply derived from the common methods of that trait.
- `program_instructions_iter` could be useful as a member fn, and **does** in fact rely on sanitized input. My plan right now is to only add "extensions" if they are necessary or would make things convenient. So far I haven't needed this method (since there's no actual uses of TransactionView) so haven't added it yet.